gdx/src/com/badlogic/gdx/utils/ObjectMap.java

```diff
@@ -1,846 +1,782 @@
+		initialCapacity = MathUtils.nextPowerOfTwo((int)Math.ceil(Math.max(1, initialCapacity) / loadFactor));
+		if (initialCapacity > 1 << 30)
+			throw new IllegalArgumentException("initialCapacity is too large: " + initialCapacity);
+
+		this.loadFactor = loadFactor;
+
+		threshold = (int)(initialCapacity * loadFactor);
+		mask = initialCapacity - 1;
+		shift = Long.numberOfLeadingZeros(mask);
+
+		keyTable = (K[])new Object[initialCapacity];
+		valueTable = (V[])new Object[initialCapacity];
+	}
+
+	/**
+	 * Creates a new map identical to the specified map.
+	 */
+	public ObjectMap (ObjectMap<? extends K, ? extends V> map) {
+		this((int)Math.floor(map.keyTable.length * map.loadFactor), map.loadFactor);
+		System.arraycopy(map.keyTable, 0, keyTable, 0, map.keyTable.length);
+		System.arraycopy(map.valueTable, 0, valueTable, 0, map.valueTable.length);
+		size = map.size;
+	}
+
+	/**
+	 * Finds an array index between 0 and {@link #mask}, both inclusive, corresponding to the hash code of {@code item}.
+	 * By default, this uses "Fibonacci Hashing" on the {@link Object#hashCode()} of {@code item}; this multiplies
+	 * {@code item.hashCode()} by a long constant (2 to the 64, divided by the golden ratio) and shifts the high-quality
+	 * uppermost bits into the lowest positions so they can be used as array indices. The multiplication by a long may
+	 * be somewhat slow on GWT, but it will be correct across all platforms and won't lose precision. Using Fibonacci
+	 * Hashing allows even very poor hashCode() implementations, such as those that only differ in their upper bits, to
+	 * work in a hash table without heavy collision rates. It has known problems when all or most hashCode()s are
+	 * multiples of larger Fibonacci numbers; see <a href="https://probablydance.com/2018/06/16/fibonacci-hashing-the-optimization-that-the-world-forgot-or-a-better-alternative-to-integer-modulo/">this blog post by Malte Skarupke</a>
+	 * for more details. In the unlikely event that most of your hashCode()s are Fibonacci numbers, you can subclass
+	 * this to change this method, which is a one-liner in this form:
+	 * {@code return (int) (item.hashCode() * 0x9E3779B97F4A7C15L >>> shift);}
+	 * <br>
+	 * This can be overridden by subclasses, which you may want to do if your key type needs special consideration for
+	 * its hash (such as if you can't modify or extend a particular class that has an incorrect hashCode()). Subclasses
+	 * that don't need the collision decrease of Fibonacci Hashing (assuming the key class has a good hashCode()) may do
+	 * fine with a simple implementation:
+	 * {@code return (item.hashCode() & mask);}
+	 *
+	 * @param item a key that this method will hash, by default by calling {@link Object#hashCode()} on it; non-null
+	 * @return an int between 0 and {@link #mask}, both inclusive
+	 */
+	protected int place (final K item) {
+		// shift is always greater than 32, less than 64
+		return (int)(item.hashCode() * 0x9E3779B97F4A7C15L >>> shift);
```

Comment:
HashMap also uses a supplemental hash function (and special cases string keys). How does that compare to this?

I wonder how cuckoo ObjectMap behavior changes if this is used for one of its hash functions? It will still have the cache miss disadvantage of course.

Reply:
The main advantage here comes from that it actually does linear probing instead of just failing when the stash is full of collisions. Using this in the cuckoo-hashing framework wouldn't help much without significant hacking on its response to failure states, and in all the attempts I made I couldn't balance the need to resize capacity (to enlarge the mask) vs. resizing the stash (which gets quite slow if it can grow without check).

Also, remove() is twice as fast now.

Reply:
The problem with our cuckoo map is likely to do with the choice of hash functions. With 3 hash functions and a stash, cuckoo is not supposed to fail as it does.

You didn't address how your rehashing compares to HashMap's. It would be helpful to have a good understanding of which is better in theory.

As previously mentioned, cuckoo remove should not be slow. The work needed is the same as get, with the addition of nulling the value. The algorithms are different, cuckoo has cache misses and linear probing has variable worst case time, but I would still expect linear probing remove to be slower on average than cuckoo remove, certainly not twice as fast, because values need to be shifted. Cuckoo should be slower on put and faster for remove, get, contains. That we don't see that in the benchmarks indicates something is wrong, somewhere.

Reply:
The main issue seems to be how it handles collisions and resizing the stash.

When collisions occur, it doubles the entire capacity to add a small amount of stash slots. Better hashing can't fix this, especially if you are adding user provided strings which can all have the exact same hashcode.

As for trying to balance resizing the stash or resizing the capacity, it's hard to know which should be done. You want to resize capacity if the hashing is the problem, but stash if rehashing provides minimal benefit.

Reply:
The cuckoo issue is failure for the pathological case where all the keys have the same hash value. The stash reduces how often resizes are needed but isn't intended as a solution for bad keys. I think replacing it with linear probing is reasonable, though I still find cuckoo interesting. It's kind of funny it has constant worst case time but then doesn't have a solution for bad keys.

Other maps do poorly with bad keys but don't completely fail: chaining becomes a linked list (ouch!), linear probing does a whooole lot of comparisons.